File: leetcode-algorithms/023. Merge k Sorted List/Solution.cpp
// Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
//
// Example:
//
// Input:
// [
//   1->4->5,
//   1->3->4,
//   2->6
// ]
// Output: 1->1->2->3->4->4->5->6
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class cmp {
public:
    bool operator()(ListNode* a, ListNode* b) {
        return a->val > b->val;
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, cmp> que;

        for (auto list : lists) {
            if (list != nullptr) que.push(list);
        }

        ListNode* head = new ListNode(0);
        ListNode* p = head;
        while (!que.empty()) {
            auto top = que.top();
            que.pop();
            ListNode* newnode = new ListNode(top->val);
            head->next = newnode;
            head = newnode;
            if (top->next != nullptr) que.push(top->next);
        }

        return p->next;
    }
};

File: leetcode-algorithms/023. Merge k Sorted List/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
    	PriorityQueue<ListNode> que = new PriorityQueue<>((l1, l2) -> l1.val - l2.val);

    	for (ListNode list : lists) {
    		if (list != null) que.add(list);
    	}

    	ListNode dummy = new ListNode(1);
    	ListNode cp = dummy;

    	while (!que.isEmpty()) {
    		ListNode top = que.poll();

    		dummy.next = top;
    		dummy = top;
    		if (top.next != null) que.add(top.next);
    	}

        return cp.next;
    }
}
